pkg/cloud/services/compute/instance_test.go

```diff
@@ -64,3 +71,298 @@ func Test_getPortName(t *testing.T) {
+			},
+			want: []infrav1.Network{
+				{ID: networkCUUID, Subnet: &infrav1.Subnet{ID: subnetC1UUID}},
+				{ID: networkAUUID, Subnet: &infrav1.Subnet{ID: subnetA1UUID}},
+				{ID: networkAUUID, Subnet: &infrav1.Subnet{ID: subnetA2UUID}},
+				{ID: networkBUUID, Subnet: &infrav1.Subnet{ID: subnetB1UUID}},
+			},
+			expect: func(m *mock_networking.MockNetworkClientMockRecorder) {
+				// List tagged networks (A & B)
+				m.ListNetwork(&testNetworkListOpts).
+					Return([]networks.Network{testNetworkA, testNetworkB}, nil)
+
+				// List tagged subnets in network A (A1 & A2)
+				networkAFilter := testSubnetListOpts
+				networkAFilter.NetworkID = networkAUUID
+				m.ListSubnet(&networkAFilter).
+					Return([]subnets.Subnet{testSubnetA1, testSubnetA2}, nil)
+
+				// List tagged subnets in network B (B1)
+				networkBFilter := testSubnetListOpts
+				networkBFilter.NetworkID = networkBUUID
+				m.ListSubnet(&networkBFilter).
+					Return([]subnets.Subnet{testSubnetB1}, nil)
+			},
+			wantErr: false,
+		},
+		{
+			// Expect an error if a network filter doesn't match any networks
+			name: "Network filter matches no networks",
+			networkParams: []infrav1.NetworkParam{
+				{Filter: testNetworkFilter},
+			},
+			want: nil,
+			expect: func(m *mock_networking.MockNetworkClientMockRecorder) {
+				// List tagged networks (none for this test)
+				m.ListNetwork(&testNetworkListOpts).Return([]networks.Network{}, nil)
+			},
+			wantErr: true,
+		},
+		{
+			// Expect an error if a subnet filter doesn't match any subnets
+			name: "Subnet filter matches no subnets",
+			networkParams: []infrav1.NetworkParam{
+				{
+					UUID: networkAUUID,
+					Subnets: []infrav1.SubnetParam{
+						{Filter: testSubnetFilter},
+					},
+				},
+			},
+			want: nil,
+			expect: func(m *mock_networking.MockNetworkClientMockRecorder) {
+				// List tagged subnets in network A
+				networkAFilter := testSubnetListOpts
+				networkAFilter.NetworkID = networkAUUID
+				m.ListSubnet(&networkAFilter).Return([]subnets.Subnet{}, nil)
+			},
+			wantErr: true,
+		},
+		{
+			name: "Subnet UUID without network",
+			networkParams: []infrav1.NetworkParam{
+				{Subnets: []infrav1.SubnetParam{
+					{UUID: subnetA1UUID},
+				}},
+			},
+			want: nil,
+			expect: func(m *mock_networking.MockNetworkClientMockRecorder) {
+			},
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mockCtrl := gomock.NewController(t)
+			mockNetworkClient := mock_networking.NewMockNetworkClient(mockCtrl)
+			tt.expect(mockNetworkClient.EXPECT())
+
+			networkingService := networking.NewTestService(
+				"", mockNetworkClient, logr.Discard(),
+			)
+			s := &Service{
+				networkingService: networkingService,
+			}
+
+			got, err := s.getServerNetworks(tt.networkParams)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Service.getServerNetworks() error = %+v, wantErr %+v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
```

Comment:
Minor point: I've noticed the tests don't use Gomega for pattern matching.

The Cluster API book suggests we should use Gomega for testing (https://cluster-api.sigs.k8s.io/developer/testing.html#gomega): 

> In Cluster API all the test MUST use Gomega assertions.

You get a bit nicer output on errors if you use Gomega. An example of how to use Gomega for matching test assertions is here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/6aa74945653aefffc76528d96fff41fe19c66e9b/pkg/cloud/services/networking/floatingip_test.go#L83-L84

Reply:
As it's trivial and MUST is in caps I updated it.